### PR TITLE
Specification for service detection modules

### PR DESCRIPTION
**Pull request by TimothyAsirJeyasing**
tendrl-bug-id:Tendrl/specifications#46
Signed-off-by: Timothy Asir J <tjeyasin@redhat.com>

### CODE REVIEW

**Review by r0h4n — COMMENTED** [review]
Service "detection" you mean?

**Comment by TimothyAsirJeyasing** [comment]
Sorry, typo error :) , i will rename to Service detection

File: specs/service_deduction.adoc [hunk 1]
```diff
@@ -0,0 +1,174 @@
+// vim: tw=79
+
+= Service Deduction Module for Node Agent
```
Comment (cybernth) [inline specs/service_deduction.adoc]: Deduction?
Have seen this at multiple places in the doc. Find and Replaceall

File: specs/service_deduction.adoc [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+== Problem description
+
+This specification describes service deduction module to manage the services
+present on the nodes and populate the details; to and from the central store.
+The list could then be presented to the administrator via the API and the UI
+to allow selection of the nodes for monitoring, service operation and
+troubleshoting.
+
+== Use Cases
```
Comment (cybernth) [inline specs/service_deduction.adoc]: Usecases needs to be rewritten. To me it is as simple as getting the details  intrerested services like ceph/gluster related services, centreal store related services etc, store them in etcd and present it through an API.

File: specs/service_deduction.adoc [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+* Use Case3: During import
+  ** The list of available nodes presented to the administrator via APIs
+  ** User selects one node in the cluster
+  ** Use the selected node to extract the details of the cluster which
+     uses deducted service status
+
+== Proposed change
+
+Node Agents related changes:
+* Add a module which fetches the service details like service-name, execution
```
Comment (cybernth) [inline specs/service_deduction.adoc]: What are the services which are of interest? Can you list down?

File: specs/service_deduction.adoc [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+* Deduct the services details from the tendrl-service-module and update
+  on any changes into centerl store.
+
+* Create an object in tendrl_definitions_node_agent for service details.
+
+=== Alternatives
+None
+
+=== Data model impact:
+* Add service details in centrl store with keys
```
Comment (cybernth) [inline specs/service_deduction.adoc]: I don't think this is the right way to do this

File: specs/service_deduction.adoc [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+==== Tendrl API impact:
+
+Service status flow requires the below APIs to be exposed:
+
+* API to Get the service details
+
+----
+
+GET /1.0/GetServices
```
Comment (cybernth) [inline specs/service_deduction.adoc]: I don't think you need a top level APi for this? Can we embed these details as part of GetNodeList itself?

File: specs/service_deduction.adoc [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+=== Impacted Modules:
+
+==== Tendrl API impact:
+
+Service status flow requires the below APIs to be exposed:
+
+* API to Get the service details
+
+----
```
Comment (cybernth) [inline specs/service_deduction.adoc]: why?

File: specs/service_deduction.adoc [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+==== Notifications/Monitoring impact:
+Can consume the api to get service details, can raise a seperate spec
+for that.
+
+==== Tendrl/common impact:
+None
+
+==== Tendrl/node_agent impact:
+* Add a module to collects the service details like service-name, execution
```
Comment (cybernth) [inline specs/service_deduction.adoc]: Require Tendrl definitions - objects, flows and atoms

File: specs/service_deduction.adoc [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+=== Assignee(s):
+
+Primary assignee:
+    tjeyasin@redhat.com
+
+=== Work Items:
+TODO
+
+== Dependencies:
+TODO
```
Comment (cybernth) [inline specs/service_deduction.adoc]: Raise github issues on the impacted components and link it here

**Review by cybernth — CHANGES_REQUESTED** [review]

File: specs/service_deduction.adoc (hunk 1 above)
Comment (TimothyAsir) [reply specs/service_deduction.adoc]: Sure

File: specs/service_deduction.adoc (hunk 3 above)
Comment (TimothyAsir) [reply specs/service_deduction.adoc]: Services or Service details?

Can detect various services already present on the nodes. This includes, but is not limited to:
    ceph services
    gluster services
    tendrl api
    tendrl monitoring application
    tendrl central store

Comment (shtripat) [reply specs/service_deduction.adoc]: The services should be listed in document

Comment (TimothyAsir) [reply specs/service_deduction.adoc]: Sure

File: specs/service_deduction.adoc (hunk 5 above)
Comment (TimothyAsir) [reply specs/service_deduction.adoc]: I am ok to keep along with node list.
However, the spec suggest to have this as a separate module.
So i thought of keeping it separately to provide service related details for the service name given.
Need your opinion?

File: specs/service_deduction.adoc (hunk 8 above)
Comment (TimothyAsir) [reply specs/service_deduction.adoc]: Sure

File: specs/service_deduction.adoc (hunk 3 above)
Comment (TimothyAsir) [reply specs/service_deduction.adoc]: @anmolbabu, Please let me know the list of services required for monitoring

File: specs/service_deduction.adoc (hunk 2 above)
Comment (TimothyAsir) [reply specs/service_deduction.adoc]: Sure

File: specs/service_deduction.adoc (hunk 7 above)
Comment (TimothyAsir) [reply specs/service_deduction.adoc]: done

File: specs/service_detection.adoc [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+The list could then be presented to the administrator via the API and the UI
+to allow selection of the nodes for monitoring, service operation and
+troubleshoting.
+
+== Use Cases
+
+* Tendrl detects services like ceph/gluster and its related services details
+  of the storage node and store them in etcd and present it through an API.
+
+  ** User selects a service of a purticular node
```
Comment (shtripat) [inline specs/service_detection.adoc]: s/purticular/particular/g

File: specs/service_detection.adoc [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+memory it consumes, number of io counts on read and write, open files,
+number of threads for the ceph and gluster related services.
+
+* Create a new class for service details in persistence module and
+  add required functions and update all necessary changes in the node-agent
+  manager class.
+
+* Create a function to get the service list from the configuration
+  or to provide a default services which are related to ceph or gluster
+  serevices like glusterd, sshd, tendrl-node-agent, tendrl-api,
```
Comment (shtripat) [inline specs/service_detection.adoc]: s/serevices/services/g

File: specs/service_detection.adoc [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+None
+
+=== Data model impact:
+* Add services details in centrl store with key nodes/node_id/Services
+/service_name.
+
+=== Impacted Modules:
+
+==== Tendrl API impact:
+None. Service details will be displayed as part of node list.
```
Comment (shtripat) [inline specs/service_detection.adoc]: Shouldn't we have explicit service listing apis `/nodes/{id}/services/{service-name}`

File: specs/service_detection.adoc [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+==== Notifications/Monitoring impact:
+Can consume the api to get service details, can raise a seperate spec
+for that.
+
+==== Tendrl/common impact:
+None
+
+==== Tendrl/node_agent impact:
+* Add a module to collects the service details like service-name, execution
```
Comment (shtripat) [inline specs/service_detection.adoc]: s/collects/collect/g

File: specs/service_detection.adoc [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+* detect the services details from the tendrl-service-module and update
+  on any changes into centerl store.
+
+* Create an object in tendrl_definitions_node_agent for service details.
+
+following is the data models to be pushed to the central store for
+Services:
+
+{"nodename": [{
```
Comment (shtripat) [inline specs/service_detection.adoc]: I feel key should be `services`

File: specs/service_detection.adoc [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+=== Other deployer impact:
+None
+
+=== Developer impact:
+None
+
+== Implementation:
+Package python-psutil is used to find the service details.
+* Add python-psutil package as dependency.
+* Add a new module named "service_detction" to fetch service details
```
Comment (shtripat) [inline specs/service_detection.adoc]: s/detction/detection/s

**Comment by mkudlej** [comment]
Please add https://github.com/Tendrl/usmqe-tests/issues/13 as reference for testing.

File: specs/service_detection.adoc (hunk 10 above)
Comment (TimothyAsir) [reply specs/service_detection.adoc]: I will correct this

File: specs/service_detection.adoc (hunk 11 above)
Comment (TimothyAsir) [reply specs/service_detection.adoc]: Sure, Service details will be displayed as part of node list in addition to that we can provide an api for list service details as per your suggestion.

File: specs/service_detection.adoc (hunk 12 above)
Comment (TimothyAsir) [reply specs/service_detection.adoc]: Done

File: specs/service_detection.adoc (hunk 14 above)
Comment (TimothyAsir) [reply specs/service_detection.adoc]: Done

File: specs/service_detection.adoc (hunk 13 above)
Comment (TimothyAsir) [reply specs/service_detection.adoc]: I Agree

File: specs/service_detection.adoc [hunk 15]
```diff
@@ -55,7 +55,30 @@ None
...
+Service details will be displayed as part of node list. In addition to
+that, API to get service details
+
+----
+
+GET /1.0/nodes/{id}/services/{service-name}
+
+Sample Response:
+
+{
```
Comment (shtripat) [inline specs/service_detection.adoc:88]: It should return a list of services as below

```
[
    {
        "node_id": "357a4f0b-4531-6727-b708-158c3d98f312",
	"name": "httpd",
	"exe_path": "/usr/bin/",
	"status": "running",
	"cpu-use-percent": "10",
	"mem-use-percent": "39",
	"fds_count": "120",
	"threads_count": "1200",
	"open_files_count": "50"
    },
        ................
]
```

File: specs/service_detection.adoc [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+* detect the services details from the tendrl-service-module and update
+  on any changes into centerl store.
+
+* Create an object in tendrl_definitions_node_agent for service details.
+
+following is the data models to be pushed to the central store for
+Services:
+
+{"services": [{
```
Comment (shtripat) [inline specs/service_detection.adoc]: Close the sample service properly as a valid json

**Comment by TimothyAsir** [comment]
@shtripat, @anivargi , Please review

File: specs/service_detection.adoc [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+None
+
+=== Other end user impact:
+None
+
+=== Performance impact:
+None
+
+=== Other deployer impact:
+None
```
Comment (r0h4n) [inline specs/service_detection.adoc:157]: Is python-psutil available as rpm on fedora/centos7/rhel7?

**Review by r0h4n — CHANGES_REQUESTED** [review]

**Pull request closed without merging**